Math/NumberTheory/QuadraticIntegers.hs

```diff
@@ -0,0 +1,191 @@
+class KnownSign (x :: Sign) where
+  signVal :: proxy x -> Sign
+
+instance KnownSign 'Pos where
+  signVal _ = Pos
+
+instance KnownSign 'Neg where
+  signVal _ = Neg
+
+-- | Takes a quadratic integer and returns its sign as an integer
+signToInt :: forall s d . KnownSign s => QuadraticInteger s d -> Integer
+signToInt _ =
+  case signVal (Proxy :: Proxy s) of
+    Pos -> 1
+    Neg -> -1
+  
+---------------------------------------------------------------
+-- Data definitions for quadratic integers
+
+data QuadraticInteger (sign :: Sign) (d :: Nat) =
+  (:+) { rationalComp :: !Integer, irrationalComp :: !Integer } deriving (Eq)
+
+
+{- If the discriminant of a field is equivalent to 1 mod 4
+   then the basis for the ring of integers is strictly
+   larger than a + bx. Need to represent this other basis -}
+
+data BasisType = Integral | Fractional
+
+
+-- | Takes a quadratic int and returns its basis type
+basisType :: forall s d . (KnownNat d, KnownSign s)
+          => QuadraticInteger s d
+          -> BasisType
+basisType z
+  | disc `mod` 4 == 1 = Fractional
+  | otherwise         = Integral
+  where disc = discriminant z
+
+
+instance Show (QuadraticInteger s d) where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "x" else show (abs b) ++ "*x"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance forall s d . (KnownNat d, KnownSign s) => Num (QuadraticInteger s d) where
+  (+) (a :+ b) (c :+ d) = (a+c) :+ (b+d)
+
+  (*) x@(a :+ b) (c :+ d) =
+    case basisType x of
+      Integral -> (a*c) + (b*d)*(discriminant x) :+ (a*d + c*b)
+      Fractional -> (a*c + b*d*((discriminant x)-1) `div` 4) :+ (a*d + b*c + b*d)
+
+  abs = undefined
+  negate (a :+ b) = (-a) :+ (-b)
+  fromInteger n =
+    n :+ 0
+  signum _ = undefined
+
+
+---------------------------------------------------------------
+-- Standard geometric/algebraic functions on quadratic integers
+
+
+-- | Extracts the discriminant information from a quadratic integer
+discriminant :: forall s d . (KnownNat d, KnownSign s)
+             => QuadraticInteger s d
+             -> Integer
+discriminant x = sign * disc
+  where disc = toInteger $ natVal (Proxy :: Proxy d)
+        sign = signToInt x
+
+
+-- | Returns the conjugate of a quadratic integer
+conjugate :: forall s d . (KnownSign s, KnownNat d)
+          => QuadraticInteger s d
+          -> QuadraticInteger s d
+conjugate x@(a :+ b) =
+  case basisType x of
+    Integral -> a :+ -b
+    Fractional -> (a+b) :+ -b
+
+-- | Computes the norm of a quadratic integer
+norm :: forall s d . (KnownSign s, KnownNat d)
+     => QuadraticInteger s d
+     -> Integer
+norm x = let (a :+ _) = x * (conjugate x) in a
+
+trace :: forall s d . (KnownSign s, KnownNat d)
+      => QuadraticInteger s d
+      -> Integer
+trace x = let (a :+ _) = x + (conjugate x) in a      
```

Comment:
Not sure about naming: there is `Debug.Trace.trace` in `base`. Maybe we can use `spur` as old school Germans did? I remember it from my (Russian) books, but I do not know whether it is recognisable by American readers.